### DISCUSSION TITLE
Is it possible to have a cache that shares in-flight requests?

### DISCUSSION THREAD

**Discussion by njlr** [App-vNext/Polly#2002] [Q&A] [answered]
Suppose I fire off two requests in quick succession, with the same cache key, via a Polly cache. 

Is it possible for them to share execution results?

The second request would not trigger any extra execution; instead it would be passed the result of the first request once it is ready.

**Comment by peter-csala** [comment, marked as answer]
The [cache policy implements](https://github.com/App-vNext/Polly/wiki/Cache#premise-some-proportion-of-requests-may-be-similar) a [cache-aside pattern](https://learn.microsoft.com/en-us/previous-versions/msp-n-p/dn589799(v=pandp.10)?redirectedfrom=MSDN). 

![Cache aside](https://learn.microsoft.com/en-us/previous-versions/msp-n-p/images/dn589799.55b56fd8930e405c5bf9580e455a16c1(en-us,pandp.10).png)

That means if your second request does not find any item in the cache with the given key 
then it passes through the cache and fetches the data from the original data source, just like your first request.

**Reply by njlr** [reply]
It seems that to get this functionality, another library or some custom code must be written. 

Thank you!

**Reply by peter-csala** [reply]
Yes, you can write a custom code to monitor the in-flight requests on a cache key basis. If there is already a pending request then you can block the new request and wait for the result. 
The tricky part here is the handling of the unhappy paths.

**Reply by njlr** [reply]
See https://github.com/Polly-Contrib/Polly.Contrib.DuplicateRequestCollapser
